### PR TITLE
JDK-8294529 : IGV: Highlight the current graphs in the Outline

### PR DESCRIPTION
**Pull request by tobiasholenstein**
# Problem
IGV marks the graph(s) of the current active `EditorTopComponent` as selected. Here is an example of a difference graph between the `Inremental Boxing Inline` and `Before CountedLoop` graphs:
before1

The selection can be changed when the user selects another graph (without opening in). This is needed for example to selected graphs that the user wants to delete:
before2

# Proposed Solution
Change the font of the open graph to `bold` and make the icon darker. Do the same for the folder where the graph is located. Now the user can select different graphs and still see which graphs are currently opened in the active `EditorTopComponent`. 
new_highlighting

# Implementation Details
Introduce a new `selected` variable in `FolderNode` and `GraphNode`. Whenever the currently viewed graphs change, the function `changed(InputGraphProvider lastProvider)` in `OutlineTopComponent` is called. Here we set the `FolderNode` and `GraphNode` to be (un)selected. This fires a `fireDisplayNameChange` and a `fireIconChange` which updates the text/icons.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294529](https://bugs.openjdk.org/browse/JDK-8294529): IGV: Highlight the current graphs in the Outline

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [43d4f9f8](https://git.openjdk.org/jdk/pull/10468/files/43d4f9f8be28fe4505d3f891423a8818216c1e51)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10468/head:pull/10468` \
`$ git checkout pull/10468`

Update a local copy of the PR: \
`$ git checkout pull/10468` \
`$ git pull https://git.openjdk.org/jdk pull/10468/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10468`

View PR using the GUI difftool: \
`$ git pr show -t 10468`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10468.diff">https://git.openjdk.org/jdk/pull/10468.diff</a>

</details>

### CODE REVIEW

**Review by robcasloz — CHANGES_REQUESTED** [review]
This feature is useful and works generally well, the only issue I found is that group and graph HTML names are not escaped. Currently, this issue affects only group names (e.g. the first group in #10440's [test IGV graph file](https://github.com/openjdk/jdk/files/9653179/hello.zip)), but I would suggest to escape graph names as well for completeness.

**Comment by tobiasholenstein** [comment]
> This feature is useful and works generally well, the only issue I found is that group and graph HTML names are not escaped. Currently, this issue affects only group names (e.g. the first group in #10440's [test IGV graph file](https://github.com/openjdk/jdk/files/9653179/hello.zip)), but I would suggest to escape graph names as well for completeness.

Thanks you @robcasloz for the hint. I have applied and tested it now.

**Review by robcasloz — CHANGES_REQUESTED** [review]
Thanks for addressing my comment, HTML escaping seems to work well now. I just found a NPE issue when testing this PR manually:
1. Open https://github.com/openjdk/jdk/files/9653179/hello.zip
2. Open a graph in `virtual void java.lang.Object.<init>()`
3. Right click on another graph in the same group and select `Difference to current graph`
4. We get the following NPE, and the highlighting disappears: 
```
[INFO] java.lang.NullPointerException
[INFO] [catch] at com.sun.hotspot.igv.coordinator.OutlineTopComponent.lambda$changed$2(OutlineTopComponent.java:262)
[INFO] 	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
[INFO] 	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
[INFO] 	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
[INFO] 	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
[INFO] 	at java.base/java.security.AccessController.doPrivileged(Native Method)
[INFO] 	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
[INFO] 	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
[INFO] 	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```
I am not sure whether this is a regression introduced in this PR or a previous issue (in which case it should be reported and addressed separately).

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/FolderNode.java [hunk 1]
```diff
@@ -132,6 +137,21 @@ public void remove() {
         }
     }
 
+    private boolean selected = false;
```
Comment (chhagedorn) [inline src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/FolderNode.java:140]: You should move this field up to the other field declarations.

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java [hunk 2]
```diff
@@ -222,12 +225,22 @@ protected boolean requestFocusInWindow(boolean temporary) {
         return super.requestFocusInWindow(temporary);
     }
 
+    private GraphNode[] selectedGraphs = new GraphNode[0];
+    private final Set<FolderNode> selectedFolders = new HashSet<>();
```
Comment (chhagedorn) [inline src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java:229]: You should move these fields up to the other field declarations.

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/GraphNode.java [hunk 3]
```diff
@@ -59,6 +60,20 @@ public GraphNode(InputGraph graph) {
         this(graph, new InstanceContent());
     }
 
+    private boolean selected = false;
```
Comment (chhagedorn) [inline src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/GraphNode.java:63]: You should move this field up to the other field declarations.

**Review by chhagedorn — APPROVED** [review]
Works well and looks good (if the mentioned issue is not a new bug from this change).

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java (hunk 2 above)
Comment (tobiasholenstein) [reply src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java]: done

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/GraphNode.java (hunk 3 above)
Comment (tobiasholenstein) [reply src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/GraphNode.java]: done

**Comment by tobiasholenstein** [comment]
> I am not sure whether this is a regression introduced in this PR or a previous issue (in which case it should be reported and addressed separately).

thanks for catching that bug. I filed https://bugs.openjdk.org/browse/JDK-8294564 for that bug. Since it is not directly related to this PR, if will integrate and then do the fix in JDK-8294564.

**Comment by tobiasholenstein** [comment]
thanks @robcasloz and @chhagedorn for the reviews!

/integrate

**Pull request closed without merging**